src/lib/viewers/doc/DocBaseViewer.js

```diff
@@ -977,6 +970,135 @@ class DocBaseViewer extends BaseViewer {
+
+        for (let i = visiblePages.first.id; i <= visiblePages.last.id; i++) {
+            const page = document.querySelector(`#bp-page-${i}`);
+            this.zoomWrapper.appendChild(page);
+            page.classList.add('disable-textLayer');
+        }
+
+        // Determine the midpoint of our pinch event if it is not provided for us
+        const touchMidpoint =
+            event.pageX && event.pageY
+                ? [event.pageX, event.pageY]
+                : getMidpoint(
+                    event.touches[0].pageX,
+                    event.touches[0].pageY,
+                    event.touches[1].pageX,
+                    event.touches[1].pageY
+                );
+
+        // Set the scale point based on the pinch midpoint and scroll offsets
+        this.zoomWrapper.style['transform-origin'] = `${touchMidpoint[0] + this.docEl.scrollLeft}px ${this.docEl
+            .scrollTop -
+            this.zoomWrapper.offsetTop +
+            touchMidpoint[1]}px`;
+
+        // Calculate current scroll position and gesture offset
+        this.xScrollPercentage =
+            (this.docEl.scrollLeft + touchMidpoint[0]) / Math.max(firstVisiblePage.clientWidth, this.docEl.clientWidth);
+        this.yScrollPercentage = (this.docEl.scrollTop + touchMidpoint[1]) / this.docEl.firstChild.clientHeight;
+        this.yOffset = touchMidpoint[1];
+        this.xOffset = touchMidpoint[0];
+
+        this.originalDistance = getDistance(
+            event.touches[0].pageX,
+            event.touches[0].pageY,
+            event.touches[1].pageX,
+            event.touches[1].pageY
+        );
+    }
+
+    /**
+     * Updates the CSS transform zoom based on the distance of the pinch gesture.
+     *
+     * @protected
+     * @param {Event} event - object
+     * @return {void}
+     */
+    pinchToZoomChangeHandler(event) {
+        if (!this.isPinching) {
+            return;
+        }
+
+        const scale = event.scale
+            ? event.scale
+            : getDistance(
+                event.touches[0].pageX,
+                event.touches[0].pageY,
+                event.touches[1].pageX,
+                event.touches[1].pageY
+            ) / this.originalDistance;
+
+        if (scale === 1 || Math.abs(this.pinchScale - scale) < 0.01) {
+            return;
+        }
+
+        if (scale > 1) {
+            this.pinchScale = Math.min(scale, 10 / this.pdfViewer.currentScale);
+        } else if (scale < 1) {
+            this.pinchScale = Math.max(scale, 0.1 / this.pdfViewer.currentScale);
+        }
+
+        this.zoomWrapper.style.transform = `scale(${this.pinchScale}, ${this.pinchScale})`;
+    }
+
+    /**
+     * Replaces the CSS transform with a native PDF.js zoom and scrolls to maintain positioning.
+     *
+     * @protected
+     * @return {void}
+     */
+    pinchToZoomEndHandler() {
+        if (this.zoomWrapper && this.isPinching) {
```

Comment:
To remove a level of arrowness, you could do `if(!this.zoomWrapper || !this.isPinching) return`